File: apps/atlas-api/src/features/metric-showcase/presentation/http/handle-metric-showcase.ts
import type { RuntimeContext } from "../../../../core/runtime-context";
import type { Telemetry } from "../../../telemetry/application/ports/telemetry";
import { fw } from "../../../../fireweave/fw-harness";
import { resolveInstanceTargetingKey } from "../../../../fireweave/fw-providers";
import { emitAllMetricTypes } from "../../application/emit-all-metric-types";

function json(data: unknown, status = 200): Response {
  return new Response(JSON.stringify(data) + "\n", {
    status,
    headers: { "content-type": "application/json" },
  });
}

/** GET /metrics/showcase — flag-gated demo of counter, histogram, up-down, gauge. */
export async function handleMetricShowcase(
  _ctx: RuntimeContext,
  telemetry: Telemetry,
): Promise<Response> {
  // @fireweave-controlpoint metric-showcase
  const enabled = await fw.controlPoints.getBooleanValue("metric-showcase", false, {
    targetingKey: resolveInstanceTargetingKey(),
  });
  if (!enabled) {
    return json({ error: "metric-showcase disabled" }, 404);
  }
  const emitted = emitAllMetricTypes(telemetry);
  return json({
    ok: true,
    feature: "metric-showcase",
    emitted,
    kinds: ["counter", "histogram", "up_down_counter", "gauge"],
  });
}

File: apps/lantern-api/src/features/metric_showcase/presentation/http/handle_metric_showcase.py
import json

from src.features.metric_showcase.application.emit_all_metric_types import (
    emit_all_metric_types,
)
from src.features.telemetry.infrastructure.start_otel import Telemetry
from src.fireweave.fw_harness import fw_control_points
from src.fireweave.fw_providers import resolve_fw_env_name
import os
from fireweave import EvaluationContext


def handle_metric_showcase(telemetry: Telemetry) -> tuple[int, bytes]:
    # @fireweave-controlpoint metric-showcase
    enabled = fw_control_points().get_boolean_value(
        "metric-showcase",
        False,
        EvaluationContext(targeting_key="bench-instance"),
    )
    if not enabled:
        return 404, (json.dumps({"error": "metric-showcase disabled"}) + "\n").encode()
    emitted = emit_all_metric_types(telemetry)
    body = {
        "ok": True,
        "feature": "metric-showcase",
        "emitted": emitted,
        "kinds": ["counter", "histogram", "up_down_counter", "gauge"],
        "environment": resolve_fw_env_name(os.environ),
    }
    return 200, (json.dumps(body) + "\n").encode()
